feat(application): import application endpoint bindings

Aflynn50 opened the pull request:
<!--
The PR title should match: <type>(optional <scope>): <description>.

Please also ensure all commits in this PR comply with our conventional commits specification:
https://github.com/juju/juju/blob/main/docs/contributor/reference/conventional-commits.md
-->

<!-- Why this change is needed and what it does. -->

## Checklist

<!-- If an item is not applicable, use `~strikethrough~`. -->

- [ ] Code style: imports ordered, good names, simple structure, etc
- [ ] Comments saying why design decisions were made
- [ ] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

<!--

Describe steps to verify that the change works.

If you're changing any of the facades, you need to ensure that you've tested
a model migration from 3.6 to 4.0 and from 4.0 to 4.0.

The following steps are a good starting point:

 1. Bootstrap a 3.6 controller and deploy a charm.

```sh
$ juju bootstrap lxd src36
$ juju add-model moveme1
$ juju deploy juju-qa-test
```

 2. Bootstrap a 4.0 controller with the changes and migrate the model.

```sh
$ juju bootstrap lxd dst40
$ juju migrate src36:moveme1 dst40
$ juju add-unit juju-qa-test
```

 3. Verify no errors exist in the model logs for the agents. If there are
    errors, this is a bug and should be fixed before merging. The fix can
    either be applied to the 4.0 branch (preferable) or the 3.6 branch, though
    that needs to be discussed with the team.

```sh
$ juju debug-log -m dst40:controller
$ juju debug-log -m dst40:moveme1
```

    4. We also need to test a model migration from 4.0 to 4.0.

```sh
$ juju bootstrap lxd src40
$ juju add-model moveme2
$ juju deploy juju-qa-test
```

```sh
$ juju migrate src40:moveme2 dst40
$ juju add-unit juju-qa-test
```

    5. Verify that there are no errors in the controller or model logs.

```sh
$ juju debug-log -m dst40:controller
$ juju debug-log -m dst40:moveme2
```

-->

## Documentation changes

<!-- How it affects user workflow (CLI or API). -->

## Links

<!-- Link to all relevant specification, documentation, bug, issue or JIRA card. -->

<!-- Replace #19267 with an issue reference or link, otherwise remove this line. -->
**Issue:** Fixes #19267.

<!-- Add a Jira card reference or link, otherwise remove this line. -->
**Jira card:** JUJU-

hmlanigan reviewed (changes requested):
The work to insert `extra-binding` and `application_endpoint` should be done the in relation domain.

I'm debating this one.

hmlanigan reviewed (changes requested):
From QA, the space_uuid in application_endpoint should only be filled in if it's not the default. Before migration I changed the value of 1 binding to be a different space than the default. After migration all `space_uuid` columns are filled in for all endpoints. Only the one I changed should be.

gfouillet commented on domain/application/modelmigration/import_test.go:
```diff
@@ -942,6 +942,168 @@ func (s *importSuite) TestImportCharmActionsNestedMaps(c *gc.C) {
+		gomock.Any(),
+	).DoAndReturn(func(_ context.Context, _ string, args service.ImportApplicationArgs) error {
+		// Assert: Check that the endpoints are mapped to the correct space names
+		c.Assert(args.Charm.Meta().Name, gc.Equals, "prometheus")
```
**nitpick(maybetodo):** Make the test fails on this assert. If the failure is comprehensive even if it occurs inside a mock, this is a nitpick:

use the function to  extract the `args` and do check outside the function, in assert phase.

gfouillet commented on domain/application/modelmigration/import_test.go:
```diff
@@ -942,6 +942,168 @@ func (s *importSuite) TestImportCharmActionsNestedMaps(c *gc.C) {
+		gomock.Any(),
+	).DoAndReturn(func(_ context.Context, _ string, args service.ImportApplicationArgs) error {
+		// Assert: Check that the endpoints are mapped to the correct space names
+		c.Assert(args.Charm.Meta().Name, gc.Equals, "prometheus")
```
**nitpick(maybetodo):** Make the test fails on this assert. If the failure is comprehensive even if it occurs inside a mock, this is a nitpick:

use the function to  extract the `args` and do check outside the function, in assert phase.

gfouillet commented on domain/application/state/application_endpoint_test.go:
```diff
@@ -103,8 +103,9 @@ func (s *applicationEndpointStateSuite) TestInsertApplicationNoCharmRelationWith
 	// Arrange: No relation
 	db, err := s.state.DB()
 	c.Assert(err, jc.ErrorIsNil)
+	spaceName, _ := s.addSpace(c, "beta")
```
**question:** That's ugly... Why are you doing that ? Who cares of returning the uuid ? If you want it, just create another function which actually return the uuid. The purpose of this helper was actually to inline stuff and avoid clutter. This is clutter.

Aflynn50 replied:
Good point, I started doing this, realised I probably should have done it with a new function, then didn't back out. Have summoned the energy to change it now haha.

gfouillet commented on domain/application/state/application_endpoint_test.go:
```diff
@@ -365,6 +370,114 @@ func (s *applicationEndpointStateSuite) TestInsertApplicationEndpointUnknownRela
+	// Assert:
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(bindings, gc.HasLen, 5)
+	c.Assert(bindings[relationName1], gc.Equals, spaceUUID1)
```
**suggestion:** I think you can be a bit more expressive

```go
c.Assert(bindings, jc.DeepEquals, expectedBindings)
```
with 

```go
expectedBindings := map[string]string{
   relationName1: s.addSpace(c, "space1").UUID(),
   relationName2: s.addSpace(c, "space2").UUID(),
   extraName1: s.addSpace(c, "space3").UUID(),
   extraName2: s.addSpace(c, "space4").UUID(),
```

I let you as an exercise how to achieve this expressiveness. tip: it would solve my previous comment about ugliness. (through `s.addSpace(c, "somespace").Name()`)

Aflynn50 replied:
I'm not sure that I want too much cleverness in my test here, at the moment it is stupid but explicit. If I start nesting functions inside map declarations and things it means more change is required to maintain the test.

nvinuesa replied:
These should all be `c.Check()`, except `c.Assert(err, jc.ErrorIsNil)` right?

gfouillet reviewed (commented):
First pass.

QA 3.6 to 4.0 ok
QA 4.0 to 4.0 nok but I may made some mistake while updating DB. Will double check in next review.

Aflynn50 commented on domain/application/modelmigration/export_test.go:
```diff
@@ -282,3 +284,49 @@ func (s *exportApplicationSuite) TestApplicationExportExposedEndpoints(c *gc.C)
+	c.Check(app.EndpointBindings(), gc.HasLen, 3)
+	c.Check(app.EndpointBindings()[""], gc.Equals, network.AlphaSpaceId)
+	c.Check(app.EndpointBindings()["endpoint"], gc.Equals, spaceUUID)
+	c.Check(app.EndpointBindings()["misc"], gc.Equals, "")
```
Not necessarily, but I want to be conservative here. If we don't export it we can never get it back if we need it.

It is not technically necessary because you could check the names against the charm extra bindings and charm relations to see which ones were missing from the endpoint bindings struct. But on the other hand, in 3.6 these were all always exported with some assignment.

Aflynn50 replied:
On second thoughts, since this information can be reconstructed, I will just not export them.

hmlanigan replied:
All endpoints are bound, "" for space indicates that it's on the default space. If you do not export them here, where will the `application_endpoint` rows be created from on import.

Aflynn50 commented:
> From QA, the space_uuid in application_endpoint should only be filled in if it's not the default. Before migration I changed the value of 1 binding to be a different space than the default. After migration all `space_uuid` columns are filled in for all endpoints. Only the one I changed should be.

Was this testing 3.6 -> 4.0 or 4.0 -> 4.0?

If it was 3.6 to 4.0 this is the expected behvaiour. As stated in the PR description, 3.6 sets defaults by setting every endpoint to the default space explicitly so the information of weather the user explicitly set the endpoint vs set the default is lost. Because of that we have to set what is there, and assume none of the default.

If this was 4.0 -> 4.0 this is a bug, can you give more steps to reproduce it?

Aflynn50 commented on domain/application/state/application_endpoint.go:
```diff
@@ -318,3 +318,73 @@ WHERE uuid =  $setDefaultSpace.uuid`, app)
+
+	endpoints := make(map[string]string, len(dbEndpoints)+len(dbExtraEndpoints)+1)
+	for _, e := range dbEndpoints {
+		endpoints[e.EndpointName] = e.SpaceUUID
+	}
+	for _, e := range dbExtraEndpoints {
+		endpoints[e.EndpointName] = e.SpaceUUID
```
Fair, have put the comments in export.go where it sets it in the description package.

hmlanigan commented:
> If it was 3.6 to 4.0 this is the expected behavior. As stated in the PR description, 3.6 sets defaults by setting every endpoint to the default space explicitly so the information of weather the user explicitly set the endpoint vs set the default is lost. Because of that we have to set what is there, and assume none of the default.

On import you do not really know if the data is coming from a prior release of juju or the most current. 

Migration is a chance to put old relation data into the new forms. There will be no change to functionality on the user end with this change, only to the data saved.

hmlanigan commented:
@Aflynn50 87edd0f does not solve the QA problem I found, even if it didn't cause a nil pointer panic on the destination controller.  During import, all `application_endpoint` rows must be created for an application. If the endpoint is bound to the default space, the `space_uuid` column is empty. Other wise it should contain the uuid of the space specified. The same as when an application is deployed.

At the end of a migration, all of the data should be in the proper form for the destination version of juju independent of where  it came from.

Aflynn50 commented:
> @Aflynn50 [87edd0f](https://github.com/juju/juju/commit/87edd0f0bac21889ac20c1eee1ad766906c2d1cc) does not solve the QA problem I found, even if it didn't cause a nil pointer panic on the destination controller. During import, all `application_endpoint` rows must be created for an application. If the endpoint is bound to the default space, the `space_uuid` column is empty. Other wise it should contain the uuid of the space specified. The same as when an application is deployed.
> 
> At the end of a migration, all of the data should be in the proper form for the destination version of juju independent of where it came from.

As far as I can tell, and in all my testing, this is what happens. Can you give specific steps to reproduce the issue?

Aflynn50 commented:
/build

Aflynn50 commented on domain/application/modelmigration/import.go:
```diff
@@ -663,6 +669,42 @@ func (i *importOperation) importCharmActions(data description.CharmActions) (*in
+			// not included explicitly in create application will be set to the
+			// application default, so do not set it here.
+			// In migrations from 3.6 the space ID should never be empty.
+			continue
```
So I did stop migrating them, but today had a discussion with Heather and have now added them back in. The argument being that there are a few other places that the exported model is used in Juju than migration (e.g. juju_dump_model) and in 3.6 these were exported.

hmlanigan commented:
The panic comes if you migrate a model which has a relation.

hmlanigan commented on domain/application/modelmigration/import.go:
```diff
@@ -671,14 +671,33 @@ func (i *importOperation) importCharmActions(data description.CharmActions) (*in
+	var applicationDefaultSpaceID string
+	for endpoint, spaceID := range app.EndpointBindings() {
+		if endpoint == "" {
+			applicationDefaultSpaceID = spaceID
```
todo: add a `continue`, there is only one default space

Aflynn50 replied:
I've added a `break`, which I guess is what you meant :)

hmlanigan commented on domain/application/modelmigration/import.go:
```diff
@@ -671,14 +671,33 @@ func (i *importOperation) importCharmActions(data description.CharmActions) (*in
+			// We don't export the alpha space to the description model, so we
+			// verify if the space ID is either the 4.0+ alpha space ID or the
+			// legacy ID ("0").
+			endpointBindings[endpoint] = network.AlphaSpaceName
```
issue: this appears to be the line causing problems in QA. Only the `space_uuid` `019644f2-3c6c-7d4c-9ff5-2d966c3a0557` should be listed in the `application_endpoint` table.

```
epl (model-moveme)> select * from application_endpoint
uuid					application_uuid			space_uuid				charm_relation_uuid
6a02b065-4292-41bb-82bb-e4c3752126e1	d0d4762d-723f-4b00-875a-3bd4003e7189	656b4a82-e28c-53d6-a014-f0dd53417eb6	7999664a-dd9a-46f0-83a1-691a5fce6cd3
ec613490-8a52-4b8b-85a9-87ac5a040020	d0d4762d-723f-4b00-875a-3bd4003e7189	019644f2-3c6c-7d4c-9ff5-2d966c3a0557	0f7880b3-d793-41f9-86aa-b585c14dbd19
fde130c9-b998-428c-8227-76413f384ebc	d0d4762d-723f-4b00-875a-3bd4003e7189	656b4a82-e28c-53d6-a014-f0dd53417eb6	5067e630-7d9c-4979-8201-686734f6d4f9
fc5c906e-9034-4876-813b-97163fe92c59	d0d4762d-723f-4b00-875a-3bd4003e7189	656b4a82-e28c-53d6-a014-f0dd53417eb6	017c173b-f9d9-4702-8800-548bcc79f94f
854e89de-5be3-4b7e-89f6-e43acdc115f2	d0d4762d-723f-4b00-875a-3bd4003e7189	656b4a82-e28c-53d6-a014-f0dd53417eb6	926f7cc5-f99e-41eb-8139-c5679a9c4a71
40e20a27-358c-49f7-8f00-ed8341c79c7e	d0d4762d-723f-4b00-875a-3bd4003e7189	656b4a82-e28c-53d6-a014-f0dd53417eb6	f1453682-c5fb-4c3a-89c2-ebb6fba1ee17
49cfb40f-1f45-42c3-8158-9eb62f648f06	a73615a2-0b71-4db0-8295-827e930dd862	656b4a82-e28c-53d6-a014-f0dd53417eb6	0ce6ee20-50b2-4c52-8b26-f4827adb9484

repl (model-moveme)> select * from space
uuid					name
656b4a82-e28c-53d6-a014-f0dd53417eb6	alpha
019644f2-3c5e-7141-8f85-3f75ed75e0ed	beta
019644f2-3c6c-7d4c-9ff5-2d966c3a0557	test

repl (model-moveme)> select * from application
uuid					name	life_id	charm_uuid				charm_modified_version	charm_upgrade_on_error	space_uuid
d0d4762d-723f-4b00-875a-3bd4003e7189	grafana	0	d5bac014-c5ef-4494-8ac7-4561ce9ee441	<nil>			false			656b4a82-e28c-53d6-a014-f0dd53417eb6
a73615a2-0b71-4db0-8295-827e930dd862	qa	0	c0e16062-346a-4956-8d97-d8f56dd721fb	<nil>			false			656b4a82-e28c-53d6-a014-f0dd53417eb6

```

Aflynn50 replied:
This was an issue with the order of the if/else clause the alpha space ID was getting imported explicitly even if it was the application default. I've swapped the order round and added a test for this.

hmlanigan reviewed (commented):
We're going to have to come back to this and fix some bugs once 4.0->4.0 migration is healthier. My binding got lost in the migration.

Aflynn50 commented:
> We're going to have to come back to this and fix some bugs once 4.0->4.0 migration is healthier. My binding got lost in the migration.

Is this a separate issue to the one mentioned in the comments? If so, can you post what QA steps you did? I can't do much to fix it if I don't know what the issue is.

Aflynn50 commented:
> The panic comes if you migrate a model which has a relation.

This is because we haven't implemented migration for relations yet, and it is still using old state code. For this PR, you don't need to create a relation, the endpoint bindings are on the application.

gfouillet reviewed (approved):
QA OK
